Proxy site

mgiuca opened the pull request:
Added proxy site to as an intermediary between requester and handler.

Handlers using the polyfill will now prompt the user to be registered
locally upon loading. Requesters using the polyfill will now prompt the
user to pick a registered handler (rather than being required to
hard-code a specific handler).

mgiuca committed:
Added proxy, as intermediary between requester and handler.

The new "proxy" site is designed to be hosted at a fixed, public web
address, which all handlers and requesters using the polyfill will talk
to. The proxy will track all registered handlers, and provide the
chooser UI.

polyfill: The requester is no longer required to specify a hard-coded
handler URL. (Rather, the proxy now contains a hard-coded handler URL,
but this is temporary.) The polyfill loads the proxy inside the iframe,
instead of the handler.

mgiuca committed:
polyfill/README: Updated to reflect the goal of this branch.

mgiuca committed:
polyfill: Do not hide the iframe.

Currently this is hideous, but we'll make it a nice popup dialog.

mgiuca committed:
polyfill: Minimal styling to make the proxy chooser readable.

mgiuca committed:
requester/README: Update instructions.

mgiuca committed:
proxy: Hide the iframe.

mgiuca committed:
proxy: Generalize to a list of handlers (currently with only one).

mgiuca committed:
proxy: Let the user choose which app to use.

Programmatically populates the dialog with radio buttons (currently
based on a hard-coded handler list, but ostensibly with the handlers
chosen by the user). Then forwards the action onto the URL of the chosen
handler.

mgiuca committed:
polyfill: Close the chooser dialog when the user clicks Open.

mgiuca committed:
proxy: Hook up the cancel button.

polyfill: Don't throw an error if the user presses cancel.

mgiuca committed:
proxy: Include Material Design Lite scripts.

mgiuca committed:
proxy: Abstract creation of radio button.

mgiuca committed:
proxy: Populate the handler UI after receiving a port.

This isn't necessary now, but once we filter based on the action, we'll
only know what handlers to show after receiving a message, so this is
the correct place.

mgiuca committed:
proxy: Use Material Design Lite style.

Now creates a 'card' and places the radio buttons within.

mgiuca committed:
proxy: Added color to chooser title.

mgiuca committed:
proxy: Change to blue_grey-deep_purple color scheme.

mgiuca committed:
proxy: Error handling.

mgiuca committed:
proxy: Use button instead of a.

mgiuca committed:
proxy: Chooser displays a proper error message if no handlers registered.

Also disables the Open button.

mgiuca committed:
proxy: Added control panel page at index.html.

Currently just contains dummy entries and has no logic behind it.

mgiuca committed:
Added index.js.

mgiuca committed:
Added proxy/registry.js with some indexeddb code.

mgiuca committed:
Added getAllHandlers.

mgiuca committed:
Use the url as a key instead of an autoIncrement.

mgiuca committed:
Use a Handler class instead of plain Objects.

mgiuca committed:
Better documentation.

mgiuca committed:
Added deleteHandlerForUrl, getHandlerForUrl and getHandlersForVerb.

mgiuca committed:
addHandler and deleteHandlerForUrl better error handling.

Now rejects the promise on error, instead of throwing an exception.

mgiuca committed:
proxy control panel: Use a MDL styled table.

mgiuca committed:
Move delete button to a single button below that deletes all selected handlers.

mgiuca committed:
registry: Made a bunch of functions top-level public.

mgiuca committed:
proxy control panel: Dynamically generate the table rows from the database.

Also enable/disable the delete button as the selection changes.

mgiuca committed:
Remember to close the db.

mgiuca committed:
proxy control panel: Delete button now deletes selected handlers.

mgiuca committed:
Always close the db.

mgiuca committed:
proxy control panel: Implement add button.

mgiuca committed:
proxy: Upgrade to MDL 1.0.5.

Fixes add button (clearing text fields now works).

mgiuca committed:
Tooltip.

mgiuca committed:
registry: Renamed database to ProxyRegistry.

mgiuca committed:
Use MDL-style tooltips.

mgiuca committed:
proxy control panel: Better error checking and reporting.

mgiuca committed:
proxy: Split up function.

mgiuca committed:
proxy/choose: Change 'title' to 'name' for consistency.

mgiuca committed:
proxy chooser: Use the user's real database instead of a hard-coded set.

mgiuca committed:
chooser: Only show the handlers for the verb 'open'.

(It is hard-coded at the moment; it's going to be tricky to get the verb
out of the action metadata.)

mgiuca committed:
proxy: Added registration page (proxy/register).

Currently just a skeleton; doesn't work.

mgiuca committed:
polyfill: Attempt to register handlers.

When loaded in a foreground page, fetches the manifest, parses it, and
prepares to register the handler with the proxy server. At the moment,
doesn't actually send a message to the proxy server.

mgiuca committed:
polyfill: Abstract creation of popup iframe.

mgiuca committed:
polyfill: Show the registration UI when a manifest is discovered with action handlers.

mgiuca committed:
proxy/register: Fill in text and respond to button clicks.

Now correctly shows the user the registration details, and allows the
user to close the dialog. But registration is a no-op.

mgiuca committed:
proxy: Validate the data.

This prevents a malicious handler from being able to insert weird data
into the database.

mgiuca committed:
proxy/register: Now registers the handler when user clicks OK.

mgiuca committed:
proxy/register: If already registered, immediately close the dialog.

mgiuca committed:
registry: put handlers instead of adding.

Renamed addHandler to registerHandler.

Registering a handler that already existed now overwrites, rather than
rejecting.

mgiuca committed:
proxy/register: Avoid flash of card if already registered.

Now hides the card by default, so if we choose not to show the UI, the
user doesn't see a flash of the registration card.

mgiuca committed:
choose: Use the verb provided by the requester.

The "options" dict from the requester is now provided to the chooser.
Previously it just used the hard-coded verb "open".

mgiuca committed:
requester/README: Rewrote the demo instructions.

mgiuca committed:
Added proxy/README.

sammc commented on polyfill/ballista-polyfill.js:
```diff
@@ -101,10 +111,35 @@ if (self.WorkerNavigator !== undefined) {
+  });
+}
+
 // Establishes a connection with the polyfill handler (by embedding a page from
```
Update this.

mgiuca replied:
What?

mgiuca replied:
Done.

sammc commented on polyfill/ballista-polyfill.js:
```diff
@@ -114,39 +149,53 @@ function sendPortToHandler(url, port) {
-// the handler's domain in an iframe). Returns (in a promise) a MessagePort on
-// succesful connection to the handler.
-function connectToHandler(url) {
+// Prompts the user for a handler app and establish a connection with the chosen
```
establishes

mgiuca replied:
Done.

sammc commented on polyfill/ballista-polyfill.js:
```diff
@@ -114,39 +149,53 @@ function sendPortToHandler(url, port) {
         channel.port1.onmessage = null;
         resolve(channel.port1);
+      } else if (e.data.connected == false) {
+        // Fail silently. (This happens if the user cancels the dialog.)
```
Don't leave the promise hanging. Cancellation needs to be reported.

mgiuca replied:
Done.

sammc commented on polyfill/ballista-polyfill.js:
```diff
@@ -354,4 +389,53 @@ function onMessageReceived(data, port) {
+    if (e.data.close) {
+      closeChooser();
+    } else {
+      reject(new Error("Received unexpected response from proxy."));
```
There is no promise here to reject.

mgiuca replied:
Done.

mgiuca committed:
Use Promise.all instead of hand-rolled waiting on multiple promises.

mgiuca committed:
Minor.

mgiuca committed:
polyfill: Throw rather than rejecting.

There was no Promise to reject.

mgiuca committed:
Update comment.

mgiuca committed:
performAction now rejects if the user cancels (instead of dangling).

mgiuca commented:
OK responded to your comments.

Umm... yeah GitHub is pretty bad for code reviews (possibly because I rebased). You should look at the delta between    e633c2c and 06ec7ae.

mgiuca commented:
Continuing code review on https://codereview.chromium.org/1419423003/.

mgiuca committed:
Use 0-arg functions instead of 'unused'.

mgiuca committed:
Use string templates instead of +.

mgiuca committed:
chooser: Do not allow the UI to be populated multiple times.

Removed code for resetting the UI; now only works the first time.

mgiuca committed:
Update URL.

mgiuca committed:
Use textContent instead of innerText.

mgiuca committed:
Use console.debug instead of console.log.

mgiuca committed:
register: Abstract handler equality into handlersEqual function.

mgiuca committed:
handlersEqual now sorts arrays.

(Doesn't care about the order of verbs.)

mgiuca committed:
Avoid redundant 'new Promise' wrappers.

mgiuca committed:
registry: Added missing semicolons.

mgiuca committed:
registry: Convert to object-oriented API.

All functions that operate on the db are now methods of a RegistryDatabase class.

mgiuca committed:
registry: deleteHandlerForUrl returns a promise.

mgiuca committed:
registry: Simplify using Promise.resolve().

mgiuca committed:
registry: registerHandler and deleteHandlerForUrls now returns undefined.

Previously returned a useless array of undefineds.
